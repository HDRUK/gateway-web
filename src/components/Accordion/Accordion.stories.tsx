import type { Meta, StoryObj } from "@storybook/react";
import { useState } from "react";
import Accordion, { AccordionProps } from "./Accordion";

const meta: Meta<typeof Accordion> = {
    component: Accordion,
};

export default meta;

type Story = StoryObj<typeof Accordion>;

const DummyComponent = () => {
    const [expanded, setExpanded] = useState<string | false>(false);
    const accordionData = [
        {
            id: "panel1",
            heading: "Accordion 1",
            details: "Accordion 1 details",
        },
        {
            id: "panel2",
            heading: "Accordion 2",
            details: "Accordion 2 details",
        },
        {
            id: "panel3",
            heading: "Accordion 3",
            details: "Accordion 3 details",
        },
    ];

    const handleChange = (isExpanded: boolean, panel: string) => {
        setExpanded(isExpanded ? panel : false);
    };

    return (
        <div>
            {accordionData.map(
                (accordion: {
                    id: string;
                    heading: string;
                    details: string;
                }) => (
                    <Accordion
                        expanded={expanded === accordion.id}
                        heading={accordion.heading}
                        contents={accordion.details}
                        onChange={(event, isExpanded) =>
                            handleChange(isExpanded, accordion.id)
                        }
                    />
                )
            )}
        </div>
    );
};

const Heading = () => <div>React Element as Heading</div>;
const Content = () => <div>React Element as Content</div>;

export const SimpleAccordion: Story = {
    render: (props: AccordionProps) => <Accordion {...props} />,
};

export const AccordionGroup: Story = {
    render: () => <DummyComponent />,
};

export const ReactElementAccordion: Story = {
    render: () => <Accordion heading={<Heading />} contents={<Content />} />,
};
