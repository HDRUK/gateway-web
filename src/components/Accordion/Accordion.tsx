import MuiAccordion, {
    AccordionProps as MuiAccordionProps,
} from "@mui/material/Accordion";
import MuiAccordionSummary from "@mui/material/AccordionSummary";
import MuiAccordionDetails from "@mui/material/AccordionDetails";
import Typography from "@mui/material/Typography";
import ArrowDropDownIcon from "@mui/icons-material/ArrowDropDown";
import { ReactNode } from "react";

export interface AccordionProps extends Omit<MuiAccordionProps, "children"> {
    heading: string | ReactNode;
    contents: string | ReactNode;
}

const Accordion = ({ heading, contents, ...rest }: AccordionProps) => {
    return (
        <div>
            <MuiAccordion {...rest}>
                <MuiAccordionSummary
                    expandIcon={<ArrowDropDownIcon color="primary" />}>
                    {heading}
                </MuiAccordionSummary>
                <MuiAccordionDetails>
                    <Typography>{contents}</Typography>
                </MuiAccordionDetails>
            </MuiAccordion>
        </div>
    );
};

export default Accordion;
