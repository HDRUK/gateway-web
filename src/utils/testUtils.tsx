import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
