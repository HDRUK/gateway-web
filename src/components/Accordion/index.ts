import Accordion from "./Accordion";

export default Accordion;

export type * from "./Accordion";
