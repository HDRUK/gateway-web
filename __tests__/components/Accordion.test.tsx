import Accordion from "@/components/Accordion";
import { render, screen } from "../testUtils";

describe("Accordion", () => {
    const heading = "Accordion Summary";
    const contents = "Detail Contents";

    it("renders without crashing", () => {
        render(<Accordion heading={heading} contents={contents} />);
    });

    it("displays the accordion summary correctly", () => {
        render(<Accordion heading={heading} contents={contents} />);

        expect(screen.getByText(heading)).toBeInTheDocument();
    });

    it("displays the detail contents correctly", () => {
        render(<Accordion heading={heading} contents={contents} />);

        expect(screen.getByText(contents)).toBeInTheDocument();
    });
    it("should render component", async () => {
        const wrapper = render(
            <Accordion heading={heading} contents={contents} />
        );

        expect(wrapper.container).toMatchSnapshot();
    });
});
