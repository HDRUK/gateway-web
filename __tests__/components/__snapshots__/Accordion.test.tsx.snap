// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render component 1`] = `
<div>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters mui-style-1elwnq4-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters mui-style-16puya-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters mui-style-o4b71y-MuiAccordionSummary-content"
        >
          Accordion Summary
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper mui-style-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium mui-style-5xmdy0-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m7 10 5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden mui-style-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical mui-style-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical mui-style-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root mui-style-15v22id-MuiAccordionDetails-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 mui-style-qcxu8i-MuiTypography-root"
                >
                  Detail Contents
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
